include/boost/geometry/algorithms/detail/overlay/graph/traverse_graph.hpp

```diff
@@ -178,73 +178,89 @@ struct traverse_graph
     bool continue_traverse(Ring& ring,
             signed_size_type component_id,
             signed_size_type start_node_id,
-            signed_size_type current_node_id)
+            signed_size_type target_node_id)
     {
-        auto const current_turn_indices = get_turn_indices_by_node_id(m_turns, m_clusters,
+        signed_size_type current_node_id = target_node_id;
+
+        std::size_t iteration_count = 0;
+
+        // Keep traversing until it finds the start (successful finish), or it is stuck,
+        // or it find an already visited node during traversal.
+        // The iteration count is a defensive check to prevent endless loops and not iterate
+        // more than times there are turns (this should not happen).
+        while (iteration_count < m_turns.size())
+        {
+            auto const current_turn_indices = get_turn_indices_by_node_id(m_turns, m_clusters,
```

Comment:
Code within this while-loop is all the same as before.
But we don't call itself recursively anymore (for large geometries on some operating systems, this could result in a stack overflow). It is just a while loop now which was a trivial change, and much better. If there are many turns (for example for equal geometries), it can loop for (for example) more than 3000 times which should not cause any problem.